server/bin/pbench-server-put-shim-001.py

```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/env python3
+# -*- mode: python -*-
+
+# This script is used to send the tarballs that a version 002 client
+# submits for further processing. It forwards the tarballs and their MD5
+# sums to the remote server using the server's HTTP PUT method
+
+import glob
+import os
+import sys
+import tempfile
+
+from collections import namedtuple
+from logging import Logger
+from pathlib import Path
+
+from copy_result_tb import CopyResultTb
+from pbench.common.exceptions import BadConfig
+from pbench.common.logger import get_pbench_logger
+from pbench.common.utils import md5sum
+from pbench.server import PbenchServerConfig
+from pbench.server.database import init_db
+from pbench.server.report import Report
+
+
+_NAME_ = "pbench-server-put-shim-001"
+
+Results = namedtuple("Results", ["nstatus", "ntotal", "ntbs"])
+
+
+def get_receive_dir(config: PbenchServerConfig, logger: Logger) -> Path:
+
+    receive_dir_prefix = config.get("pbench-server", "pbench-receive-dir-prefix")
+    if not receive_dir_prefix:
+        logger.error(
+            "Failed: No value for config option pbench-receive-dir-prefix in section pbench-server"
+        )
+        return None
+
+    receive_dir = Path(f"{receive_dir_prefix}-002").resolve()
+    if not receive_dir.is_dir():
+        logger.error("Failed: {} does not exist, or is not a directory", receive_dir)
+        return None
+
+    return receive_dir
+
+
+def process_tb(
+    config: PbenchServerConfig, logger: Logger, receive_dir: Path
+) -> Results:
+    """process_tb - receives a tar ball and copies it to the remote server
+                    using the server's HTTP PUT method
+
+        Args -
+            config -- PbenchServer config object
+            logger -- logger object to use when emitting log entries during
+                      operation
+            receive_dir -- directory where tarballs are received from agent
+
+    """
+
+    # Check for results that are ready for processing: version 002 agents
+    # upload the MD5 file as xxx.md5.check and they rename it to xxx.md5
+    # after they are done with MD5 checking so that's what we look for.
+    list_check = glob.glob(
+        os.path.join(receive_dir, "**", "*.tar.xz.md5"), recursive=True
+    )
+
+    logger.info("{}", config.TS)
+    list_check.sort()
+    nstatus = ""
+
+    ntotal = ntbs = 0
+
+    token = config.get("pbench-server", "put-token")
+
+    for tbmd5 in list_check:
+        ntotal += 1
+
+        # Extracts full pathname of tarball from the name of hashfile
+        # by trimming .md5 suffix
+        tb = Path(tbmd5[0 : -len(".md5")])
+        tbmd5 = Path(tbmd5)
+        tbdir = tb.parent
+        controller = tbdir.name
+
+        try:
+            tarball_len, tarball_md5 = md5sum(tb)
+            crt = CopyResultTb(
```

Comment:
Not sure what CopyResultTb this is. If it is the agent-side one, there are two problems:

- packaging: the class is defined in `lib/pbench/agent/results.py` which is not AFAIK packaged for the server installation.
- argument type mismatch (sort of): the `config` argument is  an instance of `PBenchServerConfig` but the class constructor expects a `PBenchAgentConfig`. I think this is "legal" in that both of these are subclasses of `PBenchConifg` and pyhon is fine with that. But depending on the implementation (e.g. if a method of `CopyResultTb` tries to use a field (? I forget the correct python terminology) of the agent config structure which does not exist in the server config structure, there is going to be a run-time error.

I haven't traced through to see that either of these possibilities is a real problem, but I wanted to flag them as potential problems.

OTOH, if there is a new server-side module that implements a CopyResultTb class, that would need to be added to this PR.

Am I barking up the wrong tree?

Reply:
> Am I barking up the wrong tree?

Nope; we'd mentioned earlier that some of this needs to be moved from agent to common and refactored slightly to accommodate both (e.g., as you pointed out, the difference in config class types). Kinda lost track of that along the way, however ...

Reply:
There is no other server-side module that is implemented. Sorry for that change

Reply:
> make sure that the config object that is passed to the constructor always has a `[results]` section with a `server_rest_url` field, so that the `copy_result_tb` method is happy. That exists in the `pbench-agent-default.cfg` file but would need to be added to the server one.

I'm generally on-board with this plan, however, I'm at odds with the details at the end.

It is clear that `CopyResultTb()` needs to know where the server is, so, when its constructor is called, the specified configuration better have a `[results]` section with a `server_rest_url` field.  However, the typical Pbench Server deployment will never call `CopyResultTb()`, and so it should not need that field (and possibly not that section).  It's only the put shim server which needs that.

So, @ndokos, does the plan still work if we don't modify the `pbench-server-default.cfg` file in the repo?  (We would only add the `server_rest_url` field to the put shim server deployment.)